File: leetcode/435.non-overlapping-intervals.cpp
// @leet start
#include <algorithm>
#include <limits>
#include <vector>

class Solution {
 public:
  int eraseOverlapIntervals(std::vector<std::vector<int>>& intervals) {
    // Sort by ascending end time
    std::sort(intervals.begin(), intervals.end(),
              [](const auto& a, const auto& b) { return a[1] < b[1]; });

    int currentEnd = std::numeric_limits<int>::min();
    int numRemoved = 0;

    for (const auto& interval : intervals) {
      if (interval[0] >= currentEnd) {
        currentEnd = interval[1];
      } else {
        numRemoved++;
      }
    }

    return numRemoved;
  }
};
// @leet end

File: leetcode/435.non-overlapping-intervals.js
// @leet start
/**
 * @param {number[][]} intervals
 * @return {number}
 */
var eraseOverlapIntervals = function (intervals) {
  intervals.sort((a, b) => a[1] - b[1]);

  let numSkipped = 0;
  let currEnd = -Infinity;

  for (const [start, end] of intervals) {
    if (start < currEnd) {
      numSkipped++;
    } else {
      currEnd = end;
    }
  }

  return numSkipped;
};
// @leet end
